samples/debugmalloc/README.md

```diff
@@ -0,0 +1,328 @@
+- Introduction about debug malloc.
+- Demonstrate how to enable debug malloc through compiling options.
+- explanation and sample of debug malloc public apis.
+
+## Introduction
+
+Debug malloc is a tool that helps identifying memory leaks in enclaves. The
+main logic of debug malloc is as follows.
+
+- Whenever memory is allocated, debug malloc captures the callstack and
+associates it with the allocated memory. Additionally, debug malloc sets the
+contents of the allocated memory to zero.
+
+- Whenever memory is freed, it clears the contents to a fixed pattern to help
+detect access of freed memory. It also frees the associated callstack
+information.
+
+- When an enclave terminates, debug malloc prints information about leaks---
+those objects that have been allocated, but not freed.
+
+According to the user's developing experience, there are two major problems
+which the memory leaking detection can help to address:
+
+- A developer wants to track all memory allocations throughout the execution of an enclave
+  and to determine if any of these allocations were leaked.
+  We name it "global detection".
+
+- A developer wants to track memory allocations in a specific part of the enclave's
+  execution that he or she can specify at the source code level. We name it "local detection".
+
+## Global detection
+
+Debug malloc is shipped to the users as a separate library
+oelibdebugmalloc.a, then there are two options to enable the global detection.
+
+### Global default option
+
+The "USE_DEBUG_MALLOC" define controls the global default. It is set to "ON" in
+debug mode by default. Users can modify this default option according to the
+need, such as changing it to "ON" in release mode or changing it to "OFF" in
+debug mode.
+
+Please note that this option only works if you are building OE from source
+code that this option is only valid in the OE project and not exported.
+Additional, as a global option it will add the "-loedebugmalloc" option to all
+the enclave targets in the OE project. If you are not building OE from source
+or just want to apply this library to specific enclaves, please check the next
+"custom option".
+
+### Custom option
+
+Since oelibdebugmalloc.a is a separate library, users can turn on debug malloc
+per enclave while the default option is "OFF".
+
+To use debug malloc, users will have to link oelibdebugmalloc.a into their
+enclaves explicitly. This can be done by
+```
+      cmake:
+      target_link_libraries(enclave openenclave::oedebugmalloc
+              openenclave::oeenclave openenclave::oelibc)
+
+      make:
+      LDFLAGS=$(shell pkg-config oeenclave-$(COMPILER) --libs)
+      LDFLAGS_CUSTOM=$(subst -loelibc, -loedebugmalloc -loelibc, $(LDFLAGS))
+      $(CC) -o debugmallocenc debugmalloc_t.o enc.o $(LDFLAGS_CUSTOM)
+
+```
+
+Please note that oedebugmalloc must be specified in the linker line before
+`oelibcxx`, `oelibc` and `oecore` libraries to make sure the related funtions
+are loaded from oedebugmalloc.
+
```

Comment:
Since the report API is only mentioned under Local Detection, how is a report generated with global detection?  Does OE SDK somehow do it automatically?  Is it internally invoked as part of oe_terminate_enclave()?   How can the user control where the report is output (stdout or some file or whatever else)?

Reply:
Update readme file to say what happens for global detecttion.

Reply:
Yes the global detection is enabled by default if the enclave is linked against oedebugmalloc:
https://github.com/nonpolarity/openenclave/blob/5ca957a2dc17379f90ffca2db3c2a9153d517d72/enclave/core/sgx/calls.c#L417.
it is invoked as part of enclave termination, and the report destination is stdout.

Reply:
README.md needs to be updated to say so

Reply:
Updated, thanks!